src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/matrix_nms/matrix_nms_kernel_ref.cpp

```diff
@@ -0,0 +1,184 @@
+
+    auto classes_num_adj = classes_num;
+    if (new_params.background_class >= 0 && new_params.background_class < classes_num)
+        classes_num_adj = std::max(1, classes_num - 1);
+
+    auto max_boxes_per_batch = max_boxes_per_class * classes_num_adj;
+    if (new_params.keep_top_k >= 0)
+        max_boxes_per_batch = std::min(max_boxes_per_batch, new_params.keep_top_k);
+
+    const size_t box_info_num = batches_num * classes_num * max_boxes_per_class;
+
+    const size_t box_info_buffer_size = box_info_num * BOX_INFO_SIZE;
+    const size_t sel_boxes_num_buffer_size = batches_num * classes_num * sizeof(int);
+
+    kernel_data.internalBufferSizes.push_back(box_info_buffer_size);
+    kernel_data.internalBufferSizes.push_back(sel_boxes_num_buffer_size);
+    kernel_data.internalBufferDataType = Datatype::F32;
+
+    for (size_t i{}; i < kernels_num; ++i) {
+        auto entry_point = GetEntryPoint(kernelName, new_params.layerID, params, options, i);
+        auto jit_constants = GetJitConstants(new_params);
+        jit_constants.AddConstant(MakeJitConstant("MATRIX_NMS_STAGE_" + std::to_string(i), "true"));
+
+        jit_constants.AddConstant(MakeJitConstant("MAX_BOXES_PER_CLASS", max_boxes_per_class));
+        jit_constants.AddConstant(MakeJitConstant("MAX_BOXES_PER_BATCH", max_boxes_per_batch));
+        auto jit = CreateJit(kernelName, jit_constants, entry_point);
+
+        DispatchData dispatch_data = SetDefault(new_params, i);
+        auto& kernel = kernel_data.kernels[i];
+        KernelBase::CheckDispatchData(kernelName, dispatch_data, params.engineInfo.maxWorkGroupSize);
+        kernel.params.workGroups.global = dispatch_data.gws;
+        kernel.params.workGroups.local = dispatch_data.lws;
+        kernel.code.kernelString = GetKernelString(kernelName, jit, entry_point, params.engineInfo);
+
+        SetKernelArguments(new_params, kernel, i);
+    }
+
+    return {kernel_data};
+}
+
+float MatrixNmsKernelRef::GetKernelsPriority(const Params& params, const optional_params& options) const {
+    return FORCE_PRIORITY_1;
+}
+
+bool MatrixNmsKernelRef::Validate(const Params& p, const optional_params& o) const {
+    if (p.GetType() != KernelType::MATRIX_NMS || o.GetType() != KernelType::MATRIX_NMS) {
+        return false;
+    }
+
+    const matrix_nms_params& params = static_cast<const matrix_nms_params&>(p);
+    // inputs: boxes, scores, second output, third output
+    if (params.inputs.size() != 4)
+        return false;
+
+    return true;
+}
+
+JitConstants MatrixNmsKernelRef::GetJitConstants(const matrix_nms_params& params) const {
+    JitConstants jit = MakeBaseParamsJitConstants(params);
+
+    const auto& boxes = params.inputs[0];
+    switch (boxes.GetDType()) {
+    case Datatype::F32:
+        jit.AddConstant(MakeJitConstant("COORD_TYPE_4", "float4"));
+        break;
+
+    case Datatype::F16:
+        jit.AddConstant(MakeJitConstant("COORD_TYPE_4", "half4"));
```

Comment:
Same can be easily achieved in the kernel:
```
#define COORD_TYPE_4 MAKE_VECTOR_TYPE(INPUT0_TYPE, 4)
```

Though I don't mind to keep current solution